File: docs/test-cases/token-smuggling.py
# Test case: token-smuggling (LLM01:2025)
import anthropic
from flask import Flask, request, jsonify

app = Flask(__name__)
client = anthropic.Anthropic()

BLOCKED_DOMAINS = ["paypal.com", "apple.com", "google.com"]


@app.route("/summarize")
def summarize():
    user_review = request.args.get("review")
    # BUG: user input passed directly — may contain RTL override or zero-width chars
    prompt = f"Summarize this product review: {user_review}"
    resp = client.messages.create(
        model="claude-haiku-4-5-20251001",
        max_tokens=256,
        messages=[{"role": "user", "content": prompt}],
    )
    return jsonify({"summary": resp.content[0].text})


@app.route("/check-url")
def check_url():
    url = request.args.get("url")
    # BUG: homoglyph bypass — "раypal.com" (Cyrillic р) passes this check
    if any(domain in url for domain in BLOCKED_DOMAINS):
        return jsonify({"safe": False})
    return jsonify({"safe": True})

File: docs/test-cases/token-smuggling.java
// Test case: token-smuggling (LLM01:2025)
package com.example.reviews;

import java.util.List;
import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;

@RestController
@RequestMapping("/api")
public class ReviewController {

    private final AnthropicClient client = new AnthropicClient();
    private static final List<String> BLOCKED_DOMAINS =
        List.of("paypal.com", "apple.com", "google.com");

    @PostMapping("/summarize")
    public ResponseEntity<String> summarize(@RequestBody ReviewRequest req) {
        // BUG: user review concatenated into prompt with no Normalizer.normalize(Form.NFKC).
        // RTL override (\u202E) and zero-width joiners flow directly into the model.
        String prompt = "Summarize this product review: " + req.getReview();
        String summary = client.complete("claude-haiku-4-5-20251001", prompt, 256);
        return ResponseEntity.ok(summary);
    }

    @PostMapping("/translate")
    public ResponseEntity<String> translate(@RequestBody ReviewRequest req) {
        // BUG: no strip of control/formatting chars (\u200B, \u200C, \u2060, \uFEFF)
        // before building the prompt — attacker-hidden instructions survive.
        String prompt = String.format("Translate to English: %s", req.getReview());
        return ResponseEntity.ok(client.complete("claude-haiku-4-5-20251001", prompt, 512));
    }

    @GetMapping("/check-url")
    public ResponseEntity<Boolean> checkUrl(@RequestParam String url) {
        // BUG: homoglyph bypass — "раypal.com" (Cyrillic 'р', U+0440) passes this
        // blocklist because the string is compared without NFKC normalization.
        for (String domain : BLOCKED_DOMAINS) {
            if (url.contains(domain)) {
                return ResponseEntity.ok(false);
            }
        }
        return ResponseEntity.ok(true);
    }

    public static class ReviewRequest {
        private String review;
        public String getReview() { return review; }
        public void setReview(String r) { this.review = r; }
    }
}
